Add note to SSC+Server about garbage collection (#470)
wandb/docs: docs/guides/artifacts/delete-artifacts.md

The contents of the artifact remain as a soft-delete, or pending deletion state, until a regularly run garbage collection process reviews all artifacts marked for deletion. The garbage collection process deletes associated files from storage if the artifact and its associated files are not used by a previous or subsequent artifact versions. 

> The sections in this page describe how to delete specific artifact versions, how to delete an artifact collection, how to delete artifacts with and without aliases, and more. You can schedule when artifacts are deleted from W&B with TTL policies. For more information, see [Manage data retention with Artifact TTL policy](./ttl.md).

Reviewer: ```suggestion
The sections in this page describe how to delete specific artifact versions, how to delete an artifact collection, how to delete artifacts with and without aliases, and more. You can schedule when artifacts are deleted from W&B with TTL policies. For more information, see [Manage data retention with Artifact TTL policy](./ttl.md). Note that whether you schedule an artifact for deletion with a TTL policy, or delete it immediately via the SDK or UI, both result in a soft-deletion and be subject to garbage collection to be transitioned to a hard-deleted state. 
```

The contents of the artifact remain as a soft-delete, or pending deletion state, until a regularly run garbage collection process reviews all artifacts marked for deletion. The garbage collection process deletes associated files from storage if the artifact and its associated files are not used by a previous or subsequent artifact versions. 

The sections in this page describe how to delete specific artifact versions, how to delete an artifact collection, how to delete artifacts with and without aliases, and more. You can schedule when artifacts are deleted from W&B with TTL policies. For more information, see [Manage data retention with Artifact TTL policy](./ttl.md).

:::note

> Artifacts that are scheduled for deletion with a TTL policy, deleted with the W&B SDK, or deleted with the W&B App UI are first soft-deleted. Artifacts that are soft deleted undergo garbage collection before they are to marked for hard-deleted.

Reviewer: ```suggestion
Artifacts that are scheduled for deletion with a TTL policy, deleted with the W&B SDK, or deleted with the W&B App UI are first soft-deleted. Artifacts that are soft deleted undergo garbage collection before they are hard-deleted.
```

The contents of the artifact remain as a soft-delete, or pending deletion state, until a regularly run garbage collection process reviews all artifacts marked for deletion. The garbage collection process deletes associated files from storage if the artifact and its associated files are not used by a previous or subsequent artifact versions. 

> <!-- :::note

Author: Note: This is commented out. Does not show up on docs.

* Set the `GORILLA_ARTIFACT_GC_ENABLED` environment variable to true: `GORILLA_ARTIFACT_GC_ENABLED=true`
* Enable bucket versioning if you use [AWS](https://docs.aws.amazon.com/AmazonS3/latest/userguide/manage-versioning-examples.html), [GCP](https://cloud.google.com/storage/docs/object-versioning) or any other storage provider such as [Minio](https://min.io/docs/minio/linux/administration/object-management/object-versioning.html#enable-bucket-versioning). If you use Azure, [enable soft deletion](https://learn.microsoft.com/en-us/azure/storage/blobs/soft-delete-blob-overview).
  :::note
  Soft deletion in Azure is equivalent to bucket versioning in other storage providers.
  :::

The following table describes how to satisfy requirements to enable garbage collection based on your deployment type. 

The `X` indicates you must satisfy the requirement:

|                                                | Environment variable | Enable versioning | 
| -----------------------------------------------| ---------------------| ----------------- |

> | Dedicated cloud with secure storage connector  |                      | X                 |

Reviewer 2: Actually enable versioning should be X for dedicated cloud with secure storage connector too, right?

Author: It is:
<img width="544" alt="Screen Shot 2023-11-08 at 1 42 28 PM" src="https://github.com/wandb/docodile/assets/15202580/991c0fe1-a467-490e-b26d-348ce687eac8">

| Customer-managed cloud                         | X                    | X                 |
| Customer managed on-prem                       | X                    | X                 |
 

> By default, dedicated cloud with secure storage connector can not edit the `GORILLA_ARTIFACT_GC_ENABLED` environment variable. Ask your account team to enable the environment variable. 

Reviewer 2: ```suggestion
By default,  `GORILLA_ARTIFACT_GC_ENABLED` environment variable is not enabled on dedicated cloud with secure storage connector. Ask your account team to enable the environment variable, preferably before you enable versioning for your bucket. 
```

> In addition, W&B encourages dedicated cloud users (with secure storage connector) to ask their account team to enable the environment variable before you enable bucket versioning (AWS, GCP, and so on) or soft deletion (Azure).

Reviewer 2: I think we could remove this, if you accept the above suggestion,

* Set the `GORILLA_ARTIFACT_GC_ENABLED` environment variable to true: `GORILLA_ARTIFACT_GC_ENABLED=true`
* Enable bucket versioning if you use [AWS](https://docs.aws.amazon.com/AmazonS3/latest/userguide/manage-versioning-examples.html), [GCP](https://cloud.google.com/storage/docs/object-versioning) or any other storage provider such as [Minio](https://min.io/docs/minio/linux/administration/object-management/object-versioning.html#enable-bucket-versioning). If you use Azure, [enable soft deletion](https://learn.microsoft.com/en-us/azure/storage/blobs/soft-delete-blob-overview).
  :::note
  Soft deletion in Azure is equivalent to bucket versioning in other storage providers.
  :::

The following table describes how to satisfy requirements to enable garbage collection based on your deployment type. 

The `X` indicates you must satisfy the requirement:

|                                                | Environment variable    | Enable versioning | 
| -----------------------------------------------| ------------------------| ----------------- | 
| Shared cloud                                   |                         |                   | 

> | Shared cloud with secure storage connector     |                         | X                 | 

Reviewer: ```suggestion
| Shared cloud with [secure storage connector](https://docs.wandb.ai/guides/hosting/secure-storage-connector)     |                         | X                 | 
```

| Dedicated cloud                                |                         |                   | 

> | Dedicated cloud with secure storage connector  |                         | X                 | 

Reviewer: ```suggestion
| Dedicated cloud with [secure storage connector](https://docs.wandb.ai/guides/hosting/secure-storage-connector)  |                         | X                 | 
```

| Customer-managed cloud                         | X                       | X                 | 
| Customer managed on-prem                       | X                       | X                 |
 

> By default, dedicated cloud with [secure storage connector](../hosting/secure-storage-connector.md) can not edit the `GORILLA_ARTIFACT_GC_ENABLED` environment variable. 

Reviewer: I am wondering if we still need this note? Is the point here to emphasize that we need to enable it for you if on dedicated cloud?

Reviewer: I am thinking this might not be necessary to add since for both dedicated rows we say that setting the env var is not a requirement (since we will handle this for dedicated instances). So a user would not attempt to try to set it themselves.

artifact = api.artifact(artifact_name)
artifact.collection.delete()
```

> ## How to enable garbage collection based on how you host W&B 

Reviewer 2: ```suggestion
## How to enable garbage collection based on how W&B is hosted
```

You might need to take additional steps to enable garbage collection if you use W&B Server. Based on your deployment type, you might need to satisfy the following requirements:

* Set the `GORILLA_ARTIFACT_GC_ENABLED` environment variable to true: `GORILLA_ARTIFACT_GC_ENABLED=true`

> * Enable bucket versioning if you use [AWS](https://docs.aws.amazon.com/AmazonS3/latest/userguide/manage-versioning-examples.html), [GCP](https://cloud.google.com/storage/docs/object-versioning) or any other storage provider such as [Minio](https://min.io/docs/minio/linux/administration/object-management/object-versioning.html#enable-bucket-versioning). If you use Azure, [enable soft deletion](https://learn.microsoft.com/en-us/azure/storage/blobs/soft-delete-blob-overview).

Reviewer 2: Should probably indicate that there's no secure storage connector with Azure, so it doesn't apply to dedicated cloud. Maybe as a note.

Reviewer 2: Basically it would apply for Azure only with Customer managed cloud and for the deployment-level bucket (there are no team-level Azure buckets yet).

Reviewer: Gotcha - good point - @ngrayluna should we add a note in the table or at the end that if you're on Azure the row dedicated cloud with SSC is not relevant for you - dedicated cloud with SSC is not an option for Azure buckets

Author: We have a note at the top of the secure storage connector page: https://docs.wandb.ai/guides/hosting/secure-storage-connector Does that suffice? 
<img width="609" alt="Screen Shot 2023-11-08 at 11 17 27 AM" src="https://github.com/wandb/docodile/assets/15202580/9df03cb9-d371-4609-85c1-a160795d6565">

Reviewer 2: Yeah I had seen that, but I feel we should add a simple note here as well since we're talking about enabling stuff on that cloud. Just so that if anyone just reads this page and doesn't go to the W&B hosting page, they can know right here itself.

Reviewer: This is a nit but I don't think we need to note that SSC is not available for azure. If the user is already on SSC they will already know this, and if they're not they will definitely go to the SSC page to learn more about it and figure that out. this page is on deleting artifacts so it is a bit redundant to add a warning on SSC availability in here. 

But if folks feel strongly I will concede on this one :). It is a nit. cc @abhinavg6 @ngrayluna

  :::note
  Soft deletion in Azure is equivalent to bucket versioning in other storage providers.
  :::

The following table describes how to satisfy requirements to enable garbage collection based on your deployment type. The `X` indicates you must satisfy the requirement.

|                                                | Environment variable | Enable versioning | 
| -----------------------------------------------| ---------------------| ----------------- |

> | Dedicated cloud with secure storage container  |                      | X                 |

Reviewer 2: For dedicated, the customer should ask their account team to enable the environment variable, and then configure the bucket versioning / soft deletion in their bucket (only for AWS/Azure). We should reflect that, so customer doesn't think that they could somehow edit that env variable themselves.

Reviewer: @szymon-piechowicz-wandb @jlzhao27 I had understood this isn't something the user needs to do and that we by default turn on this env variable for dedicated, no? If not, I think we should enable it by default. It is going to introduce a lot of friction if they need to go to their account team and the account team has to figure out how to do this, etc. Is there a reason not to enable it by default for all dedicated cloud instances?

Author: Added: 

"By default, dedicated cloud with secure storage connector can not edit the `GORILLA_ARTIFACT_GC_ENABLED` environment variable. Ask your account team to enable the environment variable. 

In addition, W&B encourages dedicated cloud users (with secure storage connector) to ask their account team to enable the environment variable before you enable bucket versioning (AWS, GCP, and so on) or soft deletion (Azure).
"

The `X` indicates you must satisfy the requirement:

|                                                | Environment variable    | Enable versioning | 
| -----------------------------------------------| ------------------------| ----------------- | 
| Shared cloud                                   |                         |                   | 
| Shared cloud with secure storage connector     |                         | X                 | 
| Dedicated cloud                                |                         |                   | 
| Dedicated cloud with secure storage connector  |                         | X                 | 
| Customer-managed cloud                         | X                       | X                 | 

> | Customer managed on-prem                       | X                       | X                 |

Reviewer: I am not sure if this is the best place for this table - this needs to go in the same place we show people the instructions for enabling versioning and setting the environment variable. This should likely be in the section where we talk about enabling GC on different hosting types.

Reviewer: We should note in these docs - if you are on saas/shared cloud or dedicated cloud - garbage collection is enabled by default (you do not have to do anything). Unless you use SSC, in that case, learn more in this section (how to enable gc based on how you host gc section, containing the matrix)

Author: I agree that this section, nor this table,  should not exist here. It should go in a separate page in Server. E.g. [See this thread](https://weightsandbiases.slack.com/archives/C064WHTC6L9/p1699316278423919).

My thinking is that we leave it here for now and make a note to remove it .

> By default, dedicated cloud with [secure storage connector](../hosting/secure-storage-connector.md) can not edit the `GORILLA_ARTIFACT_GC_ENABLED` environment variable. Ask your account team to enable the environment variable, preferably before you enable versioning for your bucket.

Reviewer: this can be removed. we will be turning it on for those we can set the var for (we be on by default after terraform changes go through)

:::note
Secure storage connector is currently only available for Google Cloud Platform and Amazon Web Services.

> :::

Reviewer: since we are removing any note of ssc in this guide - this should be moved into where we mention enabling GC for SSC

Author: This is a to do for the future. Right now I am not creating another page for this PR
